### PR TITLE
Enhanced pgAdmin 4 with support for Workspace layouts. #7708 

### PR DESCRIPTION
**Pull request by akshay-joshi**

### CODE REVIEW

File: web/pgadmin/misc/__init__.py [hunk 1]
```diff
@@ -88,6 +87,24 @@ def register_preferences(self):
                 'preview of the theme.'
             )
         )
+        self.preference.register(
+            'user_interface', 'layout',
+            gettext("Layout"), 'options', 'workspace',
+            category_label=PREF_LABEL_USER_INTERFACE,
+            options=[{'label': 'Classic', 'value': 'classic'},
```
Comment (adityatoshniwal) [inline web/pgadmin/misc/__init__.py:94]: label can gave gettext.

File: web/pgadmin/preferences/static/js/components/PreferencesComponent.jsx [hunk 2]
```diff
@@ -533,7 +533,39 @@ export default function PreferencesComponent({ ...props }) {
     }
 
     if (_data.length > 0) {
-      save(_data, data);
+      // Check whether layout is changed from Workspace to Classic.
+      let raise_layout_warning = false;
+      for (const [key, value] of Object.entries(data.current)) {
```
Comment (adityatoshniwal) [inline web/pgadmin/preferences/static/js/components/PreferencesComponent.jsx:538]: I think we can avoid looping by checking if layout is changed in `_data`. If yes then check for the value.

File: web/pgadmin/preferences/static/js/components/PreferencesComponent.jsx [hunk 3]
```diff
@@ -546,62 +578,74 @@ export default function PreferencesComponent({ ...props }) {
...
-            pgAdmin.Browser.tree.destroy().then(
-              () => {
-                pgAdmin.Browser.Events.trigger(
-                  'pgadmin-browser:tree:destroyed', undefined, undefined
-                );
-                return true;
-              }
+      // If layout is changed then only refresh the object explorer.
+      if (layout_changed) {
+        pgAdmin.Browser.tree.destroy().then(
```
Comment (adityatoshniwal) [inline web/pgadmin/preferences/static/js/components/PreferencesComponent.jsx:589]: Why do we need to refresh the tree on layout change?

File: web/pgadmin/static/js/BrowserComponent.jsx [hunk 4]
```diff
@@ -38,6 +51,14 @@ export const processesPanelData = {
   id: BROWSER_PANELS.PROCESSES, title: gettext('Processes'), content: <Processes />, closable: true, group: 'playground'
 };
 
+export const welcomeQueryToolPanelData = {
```
Comment (adityatoshniwal) [inline web/pgadmin/static/js/BrowserComponent.jsx:54]: can be removed, not used.

File: web/pgadmin/static/js/BrowserComponent.jsx [hunk 5]
```diff
@@ -38,6 +51,14 @@ export const processesPanelData = {
...
+  id: BROWSER_PANELS.WELCOME_QUERY_TOOL, title: gettext('Welcome'), content: <WorkspaceWelcomePage mode={'Query Tool'} />, closable: false, group: 'playground'
+};
+
+export const welcomePSQLPanelData = {
```
Comment (adityatoshniwal) [inline web/pgadmin/static/js/BrowserComponent.jsx:58]: can be removed, not used.

File: web/pgadmin/static/js/BrowserComponent.jsx [hunk 6]
```diff
@@ -120,41 +186,29 @@ export default function BrowserComponent({pgAdmin}) {
   if(failed) {
     return <>Failed to load preferences</>;
   }
+  /* In case of classic UI all workspace objects should point to the
+  * the instance of the default layout.
+  */
+  if (isClassicUI && pgAdmin.Browser.docker.default_workspace) {
```
Comment (adityatoshniwal) [inline web/pgadmin/static/js/BrowserComponent.jsx:192]: This code should be moved to workspaceprovider.

File: web/pgadmin/static/js/UtilityView.jsx [hunk 7]
```diff
@@ -58,6 +59,7 @@ export default function UtilityView({dockerObj}) {
                   pgAdmin.Browser.notifier.success(data.info);
                 }
                 onClose();
+                //pgAdmin.Browser.docker.default_workspace.close(panelId);
```
Comment (adityatoshniwal) [inline web/pgadmin/static/js/UtilityView.jsx:62]: remove commented code.

File: web/pgadmin/static/js/UtilityView.jsx [hunk 8]
```diff
@@ -34,6 +34,7 @@ export default function UtilityView({dockerObj}) {
       const panelId = _.uniqueId(BROWSER_PANELS.UTILITY_DIALOG);
       const onClose = ()=>docker.current.close(panelId);
       docker.current.openDialog({
+      // pgAdmin.Browser.docker.default_workspace.openDialog({
```
Comment (adityatoshniwal) [inline web/pgadmin/static/js/UtilityView.jsx:37]: remove commented code.

File: web/pgadmin/static/js/UtilityView.jsx [hunk 9]
```diff
@@ -91,6 +93,7 @@ function UtilityViewContent({schema, treeNodeInfo, actionType, formType, onClose
   /* Node type & Node obj*/
   let nodeObj = extraData.nodeType? pgAdmin.Browser.Nodes[extraData.nodeType]: undefined;
   let itemNodeData = extraData?.itemNodeData ? itemNodeData: undefined;
+  // const onClose = ()=>pgAdmin.Browser.docker.default_workspace.close(panelId);
```
Comment (adityatoshniwal) [inline web/pgadmin/static/js/UtilityView.jsx:96]: remove commented code.

**Review by adityatoshniwal — CHANGES_REQUESTED** [review]
Based on eyeball review.

File: web/pgadmin/preferences/static/js/components/PreferencesComponent.jsx (hunk 3 above)
Comment (akshay-joshi) [reply web/pgadmin/preferences/static/js/components/PreferencesComponent.jsx:589]: This is must as we are disconnecting from all servers on Layout changes

**Pull request closed without merging**